File: android/jsbridge/src/main/java/net/kibotu/jsbridge/commands/GetInsetsCommand.kt
package net.kibotu.jsbridge.commands

import androidx.core.view.WindowInsetsCompat
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import net.kibotu.jsbridge.BridgeContextProvider
import net.kibotu.jsbridge.JavaScriptBridge
import net.kibotu.jsbridge.SafeAreaService
import net.kibotu.jsbridge.commands.bottomnavigation.BottomNavigationService
import net.kibotu.jsbridge.commands.topnavigation.TopNavigationService
import net.kibotu.jsbridge.commands.utils.BridgeResponseUtils
import org.json.JSONObject
import timber.log.Timber
import java.math.RoundingMode

/**
 * Returns current system inset values so web content can adapt its layout.
 *
 * All heights are in density-independent pixels (dp) for cross-platform
 * parity with iOS points and consistency with [SafeAreaService] CSS vars.
 *
 * `safeArea` represents the total safe area insets that content must respect,
 * including both system bars and native app chrome (top/bottom navigation).
 *
 * Web usage:
 * ```javascript
 * const insets = await jsbridge.call('insets');
 * ```
 */
class GetInsetsCommand : BridgeCommand, BridgeAware {

    override var bridge: JavaScriptBridge? = null

    override val action = "insets"

    override suspend fun handle(content: Any?): JSONObject = withContext(Dispatchers.Main) {
        try {
            val activity = BridgeContextProvider.findActivity(bridge?.context)
                ?: return@withContext BridgeResponseUtils.createErrorResponse(
                    "NO_ACTIVITY",
                    "No active activity"
                )

            val density = activity.resources.displayMetrics.density
            val decorView = activity.window.decorView
            val rootInsets = decorView.rootWindowInsets

            val result = JSONObject()

            if (rootInsets != null) {
                val insetsCompat = WindowInsetsCompat.toWindowInsetsCompat(rootInsets, decorView)
                val statusBars = insetsCompat.getInsets(WindowInsetsCompat.Type.statusBars())
                val navBars = insetsCompat.getInsets(WindowInsetsCompat.Type.navigationBars())
                val ime = insetsCompat.getInsets(WindowInsetsCompat.Type.ime())

                val statusBarDp = pxToDp(statusBars.top, density)
                val systemNavDp = pxToDp(navBars.bottom, density)
                val imeDp = pxToDp(ime.bottom, density)

                val isTopNavVisible = TopNavigationService.currentConfig().isVisible
                val isBottomBarVisible = BottomNavigationService.currentVisibility()

                val topNavDp = if (isTopNavVisible) SafeAreaService.topBarHeightDp.toDouble() else 0.0
                val bottomNavDp = if (isBottomBarVisible) SafeAreaService.bottomBarHeightDp.toDouble() else 0.0

                result.put("statusBar", JSONObject().apply {
                    put("height", statusBarDp)
                    put("visible", insetsCompat.isVisible(WindowInsetsCompat.Type.statusBars()))
                })
                result.put("systemNavigation", JSONObject().apply {
                    put("height", systemNavDp)
                    put("visible", insetsCompat.isVisible(WindowInsetsCompat.Type.navigationBars()))
                })
                result.put("keyboard", JSONObject().apply {
                    put("height", imeDp)
                    put("visible", insetsCompat.isVisible(WindowInsetsCompat.Type.ime()))
                })
                result.put("safeArea", JSONObject().apply {
                    put("top", (statusBarDp + topNavDp).roundTo2())
                    put("right", pxToDp(navBars.right, density))
                    put("bottom", (systemNavDp + bottomNavDp).roundTo2())
                    put("left", pxToDp(navBars.left, density))
                })
            } else {
                result.put("statusBar", JSONObject().apply {
                    put("height", 0)
                    put("visible", true)
                })
                result.put("systemNavigation", JSONObject().apply {
                    put("height", 0)
                    put("visible", true)
                })
                result.put("keyboard", JSONObject().apply {
                    put("height", 0)
                    put("visible", false)
                })
                result.put("safeArea", JSONObject().apply {
                    put("top", 0)
                    put("right", 0)
                    put("bottom", 0)
                    put("left", 0)
                })
            }

            Timber.i("[handle] insets=$result")
            result
        } catch (e: Exception) {
            Timber.e(e)
            BridgeResponseUtils.createErrorResponse(
                "INSETS_FAILED",
                e.message ?: "Failed to get insets"
            )
        }
    }

    private fun pxToDp(px: Int, density: Float): Double =
        (px / density).toDouble().roundTo2()

    private fun Double.roundTo2(): Double =
        toBigDecimal().setScale(2, RoundingMode.HALF_UP).toDouble()
}

File: ios/JSBridge/Sources/JSBridge/Commands/GetInsetsCommand.swift
import Foundation
import UIKit

/// Returns current system inset values so web content can adapt its layout.
///
/// All heights are in points (pt) which map 1:1 to dp on Android, ensuring
/// cross-platform parity. `safeArea` includes both system bars and native
/// app chrome (top/bottom navigation) when visible.
///
/// Falls back from the view controller's window scene to the first connected scene
/// to handle edge cases where the VC's window is not yet in the hierarchy.
///
/// `@unchecked Sendable` because the weak `bridge` ref is only accessed
/// on `@MainActor`.
public final class GetInsetsCommand: BridgeCommand, BridgeAware, @unchecked Sendable {
    public let action = "insets"

    public weak var bridge: JavaScriptBridge?

    public init() {}

    @MainActor
    public func handle(content: [String: Any]?) async throws -> [String: Any]? {
        guard let vc = bridge?.viewController else {
            throw BridgeError.internalError("No view controller")
        }

        let windowScene = vc.view.window?.windowScene
            ?? UIApplication.shared.connectedScenes
                .compactMap({ $0 as? UIWindowScene }).first

        let statusBarHeight = windowScene?.statusBarManager?.statusBarFrame.height ?? 0
        let isStatusBarHidden = windowScene?.statusBarManager?.isStatusBarHidden ?? false
        let rootSafeArea = vc.view.window?.safeAreaInsets ?? vc.view.safeAreaInsets

        let isTopNavVisible = TopNavigationService.shared.config.isVisible
        let isBottomNavVisible = BottomNavigationService.shared.config.isVisible

        let topNavHeight = isTopNavVisible ? SafeAreaService.shared.topBarHeight : 0
        let bottomNavHeight = isBottomNavVisible ? SafeAreaService.shared.bottomBarHeight : 0

        let safeTop = statusBarHeight + topNavHeight
        let safeBottom = rootSafeArea.bottom + bottomNavHeight

        return [
            "statusBar": [
                "height": round2(statusBarHeight),
                "visible": !isStatusBarHidden
            ],
            "systemNavigation": [
                "height": round2(rootSafeArea.bottom),
                "visible": rootSafeArea.bottom > 0
            ],
            "keyboard": [
                "height": 0,
                "visible": false
            ],
            "safeArea": [
                "top": round2(safeTop),
                "right": round2(rootSafeArea.right),
                "bottom": round2(safeBottom),
                "left": round2(rootSafeArea.left)
            ]
        ]
    }

    private func round2(_ value: CGFloat) -> Double {
        (Double(value) * 100).rounded() / 100
    }
}
